src/tools/non-promotion-reasons.md

```diff
@@ -111,16 +163,546 @@ one or more of the following:
+you can work around the problem by giving them different names:
+
+{:.good}
+```dart
+class Example {
+  final int? _i;
+  Example(this._i);
+}
+
+class Unrelated {
+  int? get _j => Random().nextBool() ? 1 : null;
+}
+
+f(Example x) {
+  if (x._i != null) {
+    int i = x._i; // OK
+  }
+}
+```
+
+### Conflict with non-promotable field elsewhere in library {#field-name}
+
+**The cause:**
+You're trying to promote a field, but another class in the same library contains
+a field with the same name that isn't promotable
+(for any of the other reasons listed on this page).
+
+**Example:**
+
+{:.bad}
+```dart
+class Example {
+  final int? _overridden;
+  Example(this._overridden);
+}
+
+class Override implements Example {
+  @override
+  int? _overridden; // CONTEXT
+}
+
+f(Example x) {
+  if (x._overridden != null) {
+    int i = x._overridden; // ERROR
+  }
+}
+```
+
+This example fails because at runtime, `x` might actually be an
+instance of `Override`, so promotion would not be sound.
+
+The failure could also occur between fields in
+[unrelated clasess](#note-about-unrelated-classes).
+
+**Message:**
+
+```nocode
+'overridden' couldn’t be promoted because there is a conflicting non-promotable field in class 'Override'.
+```
+
+**Solution:**
+
+*// Try renaming one of the conflicting fields? What if they ARE related?*
+
+{:.good}
+```dart
+
+```
+
+### Conflict with implicit `noSuchMethod` forwarder {#nosuchmethod}
+
+**The cause:**
+You're trying to promote a field that is private and final,
+but another class in the same library contains an
+[implicit `noSuchMethod` forwarder][nosuchmethod]
+with the same name as the field.
+
+This is unsound because there’s no guarantee that `noSuchMethod`
+will return a stable value from one invocation to the next.
+
+**Example:**
+
+{:.bad}
+```dart
+import 'package:mockito/mockito.dart';
+
+class Example {
+  final int? _i;
+  Example(this._i);
+}
+
+class MockExample extends Mock implements Example {} // CONTEXT
+
+f(Example x) {
+  if (x._i != null) {
+    int i = x._i; // ERROR
+  }
+}
+```
+
+In this example, `_i` can't be promoted because it could resolve to the unsound
+implicit `noSuchMethod` forwarder (also named `_i`) that the compiler generates
+inside `MockExample`. 
+
+The compiler creates this implicit implementation of `_i` because
+`MockExample` promises to support a getter for `_i` when it implements
+`Example` in its declaration, but doesn't fulfill that promise. 
+So, the undefined getter implementation is handled by `Mock`'s `noSuchMethod`
+definition, which creates an implicit `noSuchMethod` forwarder of the same name.
+
+The failure can also occur between fields in
+[unrelated clasess](#note-about-unrelated-classes).
+
+**Message:**
+
+```nocode
+'_i' couldn’t be promoted because there is a conflicting noSuchMethod forwarder in class 'MockExample'.
+```
+
+**Solution:**
+
+*// Is there a solution here? Maybe explicitly supporting/defining the getter in `MockExample` would fix it?*
```

Comment:
@stereotype441

Reply:
Yeah, I like that idea. We could give something like this as a solution:

```dart
import 'package:mockito/mockito.dart';

class Example {
  final int? _i;
  Example(this._i);
}

class MockExample extends Mock implements Example {
  @override
  late final int? _i;
}

f(Example x) {
  if (x._i != null) {
    int i = x._i;
  }
}
```

(I made the override `late` because that's consistent with the way mocks generally get used--first you construct them and then you tell them how they should behave.)